Dev spring security

iryna-pantsialeyeva opened the pull request:
Please review my PR
BPR-8 Add web security with "path" security

iryna-pantsialeyeva committed:
swagger friendly to spring boot 3 (jakarta) as proposal

iryna-pantsialeyeva committed:
1. update UserDTO.java:
  - add class field "password", "isActive", "authorities"
  - add implements UserDetails and realisation of UserDetails interface methods
  - add getUserDetails() and fromRolesToAuthorities()
2. update User.java:
  - add FetchType.EAGER to @ManyToMany for Set<Role> roles;
  - addRole() - because of Hashset can not be initialized with get()
  - add @Builder
3. add UserDTOForSaveUpdate.java with method for transformation UserDTOForSaveUpdate->User
4. update data.sql and import.sql:
  - add "ACTIVE" to insert into users
  - update passwords with bcrypted passwords
5. update schema.sql: add "active bit" in table users
6. update pom.xml:
  - add spring-boot-starter-security dependency
7. update UserController.java:
  - update @Authowired field declaration (with constructor)
  - User in methods' parameters replace with UserDTOForSaveUpdate, User in  methods' return type replace with  UserDTO
8. update UserService.java: add "extends UserDetailsService"
9. update UserServiceImpl.java:
   - update @Authowired fields declaration (with constructor)
   - update saveUser() and updateUser() in accordance with new methods' parameters and return type
   - add override method loadUserByUsername()
10. update UserRepository.java: update methods' return type for findByUserName()
11. add SecurityConfig.java for DaoAuthentication

iryna-pantsialeyeva committed:
1. update Role.java:
  - delete field Set<User> users;
2. update import.sql:
  - replace Dasha's roles_id with 2 (=Admin)
3. update SecurityConfig.java
  - add @EnableMethodSecurity to determine access to controler's methods
  - add 2 beans  securityFilterChain for test and dev profiles (test - to manage access to h2 without authentication, to add default formLogin)
4. update UserController.java:
  - add @PreAuthorize for class
  - limit access for deleteUser() by @PreAuthorize
5. update AppExceptionHandler: add handleAccessDeniedException()
6. add AdminController.java with methods available only for Admin
7. update UserService.java:
  - add setAdminAuthority(), banUser(), unbanUser()
8. update UserServiceImpl.java:
   - update saveUser() with adding passwordEncoder

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/config/SecurityConfig.java:
```diff
@@ -0,0 +1,61 @@
+                .defaultSuccessUrl("/swagger-ui/index.html")
+                .and()
+                .csrf().disable()
+                .headers().frameOptions().sameOrigin()
```
👏 👏 👏 
молодец!

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/controller/AdminController.java:
```diff
@@ -0,0 +1,35 @@
+
+@RestController
+@RequestMapping("/admin")
+@PreAuthorize("hasAuthority('Admin')")
```
если закрывать весь контроллер, то, наверное, лучше в конфиге безопасности

iryna-pantsialeyeva replied:
сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/controller/UserController.java:
```diff
@@ -18,9 +19,15 @@
 
 @RestController
 @RequestMapping("/users")
+@PreAuthorize("hasAnyAuthority('User', 'Admin')")
```
убери наверное отсюда

iryna-pantsialeyeva replied:
после добавления матчера админа в конфиг работает без этой аннотации. убрала

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/controller/UserController.java:
```diff
@@ -18,9 +19,15 @@
+        this.userService = userService;
+    }
 
     @GetMapping("/")
```
Вот тут можно закрыть эндпойнт только для админов (зачем юзеру знать всех остальных юзеров)?

iryna-pantsialeyeva replied:
Согласна. сделала.

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/controller/UserController.java:
```diff
@@ -33,17 +40,18 @@ public UserDTO getUserByID(@PathVariable int userID) {
-    public UserDTO updateUser(@PathVariable int userID, @RequestBody User user) {
-        return userService.updateUser(userID, user);
+    public UserDTO updateUser(@PathVariable int userID, @RequestBody UserDTOForSaveUpdate userDTO) {
+        return userService.updateUser(userID, userDTO);
```
Тут посоветую, раз уже безопасность прикручена, достать текущего юзера из контекста и сравнить айдишники - вдруг юзер хочет поменять инфу другого юзера? ))

iryna-pantsialeyeva replied:
Сделано. Спасибо за идею

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/dto/UserDTO.java:
```diff
@@ -1,22 +1,88 @@
-public class UserDTO {
+@AllArgsConstructor
+@Builder
+public class UserDTO implements UserDetails {
```
Не, UserDto - это то, что ты предоставляешь наружу. В нем не должно быть ни пароля, ли ролей. Поэтому для идентификации заведи просто отдельный класс типа BPUserDetails и храни там то, что тебе необходимо )

iryna-pantsialeyeva replied:
Сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/dto/UserDTOForSaveUpdate.java:
```diff
@@ -0,0 +1,24 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserDTOForSaveUpdate {
```
UpdateUserDto? - update - это же тот же save  по идее )) и наоборот ))

iryna-pantsialeyeva replied:
Изменила название

RomanKashenok reviewed (changes requested):
Минимально поправь пожалуйста, а так - всё хорошо! :)

iryna-pantsialeyeva committed:
1. add BPUserDetails
2. add requestMatchers for Admin path (delete annotation in AdminController)
3. rename UserDTOForSaveUpdate
4. UserDTO is not UserDetails anymore
5. update UserController:
  - exclude ability of user to getAllUsers()
  - add @AuthenticationPrincipal BPUserDetails user in updateUser()
6. add BPUserDetails user in parameters of updateUser() in UserService
7. update updateUser() in UserServiceImpl (add check id authenticated user with updating user id)
